### ISSUE TITLE
Ambiguous error reporting when trying to infer const generic type

### ISSUE THREAD

**Issue by Joeyh021** [rust-lang/rust#93750]
Given the following code: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=659888b8e5f59352118b3c7b39187f29

```rust
#![feature(adt_const_params)]
#![feature(generic_const_exprs)]
#![feature(generic_arg_infer)]

#[derive(PartialEq, Eq)]
#[allow(dead_code)]
enum Colour {
    R,
    G,
    B,
    W,
}

struct Led<const C: Colour> {}

impl<const C: Colour> Led<C> {
    pub fn new() -> Led<{ Colour::W }> {
        Led::<{ Colour::W }> {}
    }
}

fn main() {
    let l = Led::<_>::new();
}

```

The current output is:

```
error[E0284]: type annotations needed: cannot satisfy `the constant `Led::<{_: Colour}>::new::{constant#0}` can be evaluated`
  --> src\main.rs:24:13
   |
24 |     let l = Led::<_>::new();
   |             ^^^^^^^^^^^^^ cannot satisfy `the constant `Led::<{_: Colour}>::new::{constant#0}` can be evaluated`
   |
note: required by a bound in `Led::<C>::new`
  --> src\main.rs:18:25
   |
18 |     pub fn new() -> Led<{ Colour::Off }> {
   |                         ^^^^^^^^^^^^^^^ required by this bound in `Led::<C>::new`

For more information about this error, try `rustc --explain E0284`.
```

The error message being `cannot satisfy 'the constant 'Led::<{_: Colour}>::new::{constant#0}' can be evaluated'` doesn't explain the error at all, and is confusing in the way it is reported. I'm not entirely sure what is going on internally, but getting the compiler to emit a proper error message  explaining that the const generic type cannot be inferred would be useful.

**Comment by lcnr** [comment]
same issue as #92186
